src/v/cloud_storage/inventory/inv_consumer.h

```diff
@@ -0,0 +1,117 @@
+
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cloud_storage/inventory/report_parser.h"
+#include "container/fragmented_vector.h"
+#include "model/fundamental.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/iostream.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+
+namespace cloud_storage::inventory {
+
+/// The inventory consumer reads a CSV inventory report, which may be GZip
+/// compressed, and stores the hashes for file paths from the report. Each row
+/// in the report is expected to contain a path to a segment in the bucket for
+/// which the inventory was generated. A set of NTPs is provided to the consumer
+/// for validation, and only paths belonging to these NTPs are processed. For
+/// each path, its 64 bit hash is calculated and stored in memory. Once the
+/// cumulative size in memory of the stored hashes grows over a prescribed
+/// limit, the hashes for the invidividual NTPs are written to disk until the
+/// cumulative memory usage drops below the desired limit. On completion of the
+/// report consumption, all hashes in memory are flushed to disk.
+class inventory_consumer {
+    using write_all_t = ss::bool_class<struct write_all_tag>;
+
+public:
+    inventory_consumer(const inventory_consumer&) = delete;
+    inventory_consumer& operator=(const inventory_consumer&) = delete;
+    inventory_consumer(inventory_consumer&&) = delete;
+    inventory_consumer& operator=(inventory_consumer&&) = delete;
+
+    inventory_consumer(
+      std::filesystem::path hash_store_path,
+      absl::node_hash_set<model::ntp> ntps,
+      size_t max_hash_size_in_memory);
+
+    /// Consumes the report provided as input stream and populates in memory
+    /// path hashes. A flush to disk may be performed if the size of hashes
+    /// exceeds the prescribed limit.
+    ss::future<>
+    consume(ss::input_stream<char> s, gzip_compressed_t is_report_compressed);
+
+    static std::optional<model::ntp> ntp_from_path(std::string_view path);
+    static std::vector<ss::sstring> parse_row(ss::sstring row);
+
+    /// Ensures that streams opened by consumer are closed. Called during
+    /// exception handling if an exception is thrown during a stream
+    /// consumption.
+    ss::future<> stop();
+
+    ~inventory_consumer();
+
+private:
+    ss::future<> do_consume(
+      ss::input_stream<char> s, gzip_compressed_t is_report_compressed);
+
+    // Checks if the path belongs to one of the NTPs whose leadership belongs to
+    // this node. If so, the path is hashed and added to the `_path_hashes` map.
+    ss::future<> process_path(ss::sstring path);
```

Comment:
I'm assuming that leadership comes into play here as part of a strategy to distribute work across the cluster? Is there a place we can read more about the strategy? At least how I'm imagining how this works it seems error prone, from the case of no leader to the case of leadership moving at just the right times so as to have a partition which is fully ignored

Reply:
This is a by-product of how scrubber works. The scrub is done as a job started by the ntp archiver. The ntp archiver runs per partition, on the partition leader. So since the consumers of the inventory report data are scrubbers, we only need to process rows for NTPs which will have leaders on the current node, hence the filtering on basis of these NTPs. When the partition leaders on a given node run the scrub, they will expect to find the data on disk if the scrub has run, or skip the deep scrub action if the data is not present on disk.

There is a risk of partition leadership moving around while the data is being processed, but since the scrubber is going to work on a best effort basis, if while running it does not find the hashes of it's NTP on the local disk, it is expected to skip the deep scrub actions supported by inventory data (IE the specific check if each segment in manifest is present) and proceed with the high level scrub we do today. Then on the next run it can look for the inventory hashes again and use them if available. 

There isn't really a workaround for this IMO, unless we always store hashes for every partition seen in the inventory report. That could use up too much disk space when we save them to disk. Since the inventory report is generated daily, trying for a full scrub on the next run is possible every day.

I will add these details to the comment to make it clearer.

Reply:
updated the comment